misc/metrics/src/identify.rs

```diff
@@ -21,39 +21,79 @@
 use prometheus_client::metrics::counter::Counter;
-use prometheus_client::metrics::family::Family;
-use prometheus_client::metrics::histogram::{exponential_buckets, Histogram};
-use prometheus_client::metrics::MetricType;
-use prometheus_client::registry::Registry;
+use prometheus_client::metrics::family::ConstFamily;
+use prometheus_client::metrics::gauge::ConstGauge;
+use prometheus_client::registry::{Descriptor, LocalMetric, Registry};
+use prometheus_client::MaybeOwned;
+use std::borrow::Cow;
 use std::collections::HashMap;
-use std::iter;
 use std::sync::{Arc, Mutex};
 
+static PROTOCOLS_DESCRIPTOR: Lazy<Descriptor> = Lazy::new(|| {
+    Descriptor::new(
+        "remote_protocols",
+        "Number of connected nodes supporting a specific protocol, with \
+         \"unrecognized\" for each peer supporting one or more unrecognized \
+         protocols",
+        None,
+        None,
+        vec![],
+    )
+});
+static LISTEN_ADDRESSES_DESCRIPTOR: Lazy<Descriptor> = Lazy::new(|| {
+    Descriptor::new(
+        "remote_listen_addresses",
+        "Number of connected nodes advertising a specific listen address",
+        None,
+        None,
+        vec![],
+    )
+});
+static OBSERVED_ADDRESSES_DESCRIPTOR: Lazy<Descriptor> = Lazy::new(|| {
+    Descriptor::new(
+        "local_observed_addresses",
+        "Number of connected nodes observing the local node at a specific address",
+        None,
+        None,
+        vec![],
+    )
+});
+const ALLOWED_PROTOCOLS: &[StreamProtocol] = &[
+    #[cfg(feature = "dcutr")]
+    libp2p_dcutr::PROTOCOL_NAME,
+    // #[cfg(feature = "gossipsub")]
+    // #[cfg(not(target_os = "unknown"))]
+    // TODO: Add Gossipsub protocol name
+    libp2p_identify::PROTOCOL_NAME,
+    libp2p_identify::PUSH_PROTOCOL_NAME,
+    #[cfg(feature = "kad")]
+    libp2p_kad::PROTOCOL_NAME,
+    #[cfg(feature = "ping")]
+    libp2p_ping::PROTOCOL_NAME,
+    #[cfg(feature = "relay")]
+    libp2p_relay::STOP_PROTOCOL_NAME,
+    #[cfg(feature = "relay")]
+    libp2p_relay::HOP_PROTOCOL_NAME,
+];
+
 pub(crate) struct Metrics {
-    protocols: Protocols,
+    peers: Peers,
     error: Counter,
     pushed: Counter,
     received: Counter,
-    received_info_listen_addrs: Histogram,
-    received_info_protocols: Histogram,
     sent: Counter,
-    listen_addresses: Family<AddressLabels, Counter>,
 }
 
 impl Metrics {
     pub(crate) fn new(registry: &mut Registry) -> Self {
         let sub_registry = registry.sub_registry_with_prefix("identify");
 
-        let protocols = Protocols::default();
-        sub_registry.register(
-            "protocols",
-            "Number of connected nodes supporting a specific protocol, with \
-             \"unrecognized\" for each peer supporting one or more unrecognized \
-             protocols",
-            protocols.clone(),
-        );
+        let peers = Peers::default();
+        sub_registry.register_collector(Box::new(peers.clone()));
```

Comment:
I think initializing this `Metrics` struct would almost be cleaner if we would just say `Metrics::default()` and then have a separate `&self` method that takes the registry and registers each metric.

Reply:
Idea for your `prometheus-client` crate: Implementing a constructor that does this could be implemented as a derive macro that goes through all fields of the struct and registers them by:

- using the field name as a metric name
- using the rustdoc string as the metric help
- collectors would have to be tagged with an attribute to allow the macro to call the different function

Reply:
> I think initializing this `Metrics` struct would almost be cleaner if we would just say `Metrics::default()` and then have a separate `&self` method that takes the registry and registers each metric.

Ideally I would like to prevent at the type level that someone creates a `Metrics` struct but then does not register it with a `Registry`.


> Idea for your `prometheus-client` crate:

Good idea! Tracked on https://github.com/prometheus/client_rust/issues/140.